receiver/snmpreceiver/scraper.go

```diff
@@ -0,0 +1,559 @@
+	err = s.client.Close()
+
+	return
+}
+
+// scrape collects and creates OTEL metrics from a SNMP environment
+func (s *snmpScraper) scrape(_ context.Context) (pmetric.Metrics, error) {
+	// Get a basic ResourceMetrics prepped for metrics with no resource attributes
+	resourceMetrics := pmetric.NewResourceMetrics()
+	scopeMetrics := resourceMetrics.ScopeMetrics().AppendEmpty()
+	scopeMetrics.Scope().SetName("otelcol/snmpreceiver")
+	scopeMetrics.Scope().SetVersion(s.settings.BuildInfo.Version)
+	metricSlice := scopeMetrics.Metrics()
+	now := pcommon.NewTimestampFromTime(time.Now())
+
+	metricsProcessed := false
+
+	resourceMetricsMap := map[string]*pmetric.ResourceMetrics{}
+	metricsMap := map[string]*pmetric.Metric{}
+	// Try to scrape scalar OID based metrics.
+	// metricsMap is passed in as a place for any created metrics to live
+	if err := s.scrapeScalarMetrics(now, metricsMap); err != nil {
+		s.logger.Warn("Failed to collect scalar OID metrics", zap.Error(err))
+	} else if len(metricsMap) != 0 {
+		metricsProcessed = true
+		// Add new scalar metrics to the basic ResourceMetrics with no resource attributes
+		for _, value := range metricsMap {
+			value.MoveTo(metricSlice.AppendEmpty())
+		}
+		// Load the map with the basic ResourceMetrics (that possibly
+		// has scalar OID based metrics already on it)
+		resourceMetricsMap[""] = &resourceMetrics
+	}
+
+	// Load this map with the scalar metric map that was just created for the
+	// basic ResourceMetrics. The info in this map is ultimately already
+	// contained within the resourceMetricsMap, but it is more easily accessable
+	// to pull out a specific existing Metric by resource and metric name here
+	metricsByResourceMap := map[string]map[string]*pmetric.Metric{
+		"": metricsMap,
+	}
+	// Try to scrape column OID based metrics.
+	// resourceMetricsMap is passed in as a place for any created resources/metrics to live
+	// metricsByResourceMap is provided as an easy way to check if a metric is already
+	// associated with an existing resource
+	scalarMetricCnt := len(metricsMap)
+	if err := s.scrapeIndexedMetrics(now, resourceMetricsMap, metricsByResourceMap); err != nil {
+		s.logger.Warn("Failed to collect column OID metrics", zap.Error(err))
+	} else if len(metricsByResourceMap[""]) > scalarMetricCnt || len(metricsByResourceMap) > 1 {
+		metricsProcessed = true
+	}
+
+	// Return error if we failed to scrape any metrics
+	md := pmetric.NewMetrics()
+	if !metricsProcessed {
+		return md, errScrape
+	}
+
+	// Put all of the created ResourceMetrics onto a top level Metrics
+	for _, value := range resourceMetricsMap {
+		value.MoveTo(md.ResourceMetrics().AppendEmpty())
+	}
+
+	return md, nil
+}
+
+// scrapeScalarMetrics retrieves all SNMP data from scalar OIDs and turns the returned scalar data
+// into metrics with optional enum attributes
+func (s *snmpScraper) scrapeScalarMetrics(now pcommon.Timestamp, metricsMap map[string]*pmetric.Metric) (err error) {
```

Comment:
nit: Probably removed the named return value. It's only used once at the very end.

Reply:
removed